homeassistant/components/alidns/__init__.py

```diff
@@ -0,0 +1,145 @@
+"""Integrate with AliDNS."""
+import asyncio
+from datetime import timedelta
+import json
+import logging
+from re import search
+
+from aliyunsdkalidns.request.v20150109.AddDomainRecordRequest import (
+    AddDomainRecordRequest,
+)
+from aliyunsdkalidns.request.v20150109.DescribeSubDomainRecordsRequest import (
+    DescribeSubDomainRecordsRequest,
+)
+from aliyunsdkalidns.request.v20150109.UpdateDomainRecordRequest import (
+    UpdateDomainRecordRequest,
+)
+from aliyunsdkcore.acs_exception.exceptions import ClientException, ServerException
+from aliyunsdkcore.client import AcsClient
+import async_timeout
+import voluptuous as vol
+
+from homeassistant.const import CONF_DOMAIN, CONF_SCAN_INTERVAL
+import homeassistant.helpers.config_validation as cv
+
+_LOGGER = logging.getLogger(__name__)
+TIMEOUT = 30  # seconds
+INTERNET_IP_URL = "http://www.3322.org/dyndns/getip"
+
+DEFAULT_INTERVAL = timedelta(minutes=10)
+DOMAIN = "alidns"
+
+CONF_ACCESS_ID = "access_id"
+CONF_ACCESS_KEY = "access_key"
+CONF_SUB_DOMAIN = "sub_domain"
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_ACCESS_ID): cv.string,
+                vol.Required(CONF_ACCESS_KEY): cv.string,
+                vol.Required(CONF_DOMAIN): cv.string,
+                vol.Required(CONF_SUB_DOMAIN): cv.string,
+                vol.Optional(CONF_SCAN_INTERVAL, default=DEFAULT_INTERVAL): vol.All(
+                    cv.time_period, cv.positive_timedelta
+                ),
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+
+async def async_setup(hass, config):
+    """Initialize the AliDNS component."""
+    conf = config[DOMAIN]
+    access_id = conf.get(CONF_ACCESS_ID)
+    access_key = conf.get(CONF_ACCESS_KEY)
+    domain = conf.get(CONF_DOMAIN)
+    sub_domain = conf.get(CONF_SUB_DOMAIN)
+    update_interval = conf[CONF_SCAN_INTERVAL]
+
+    session = hass.helpers.aiohttp_client.async_get_clientsession()
+    acs_client = AcsClient(access_id, access_key)
+
+    result = await _update_alidns(hass, session, acs_client, domain, sub_domain)
+
+    if result is False:
+        return False
+
+    async def update_domain_callback(now):
+        """Update the AliDNS entry."""
+        await _update_alidns(hass, session, acs_client, domain, sub_domain)
+
+    hass.helpers.event.async_track_time_interval(
+        update_domain_callback, update_interval
+    )
+
+    return True
+
+
+async def _update_alidns(hass, session, acs_client, domain, sub_domain):
+    """Update AliDNS."""
+    try:
+        with async_timeout.timeout(TIMEOUT):
+            resp = await session.get(INTERNET_IP_URL)
```

Comment:
We don't allow direct calls to a requests session and similar or service specific data parsing. A 3rd party library should wrap this more.

https://developers.home-assistant.io/docs/creating_component_code_review#4-communication-with-devicesservices